### PR TITLE
Initial pass at  api to support user accounts

### PR DESCRIPTION
**Pull request by akeem**
# Description

This change exposes the ability to store user accounts and the ability to retrieve an encrypted private key based on a provided email address. 

Notes:

* We are JSON column to store the password encrypted private key. This is the least common denominator between postgres and sqllite. It might be worth a discussion around changing to the Postgres specific JSONB column type for performance resasons.

* Indices will be added in following change

# Issues

Fixes #
Refs #

# Checklist:

- [ ] 1 PR, 1 purpose: my Pull Request applies to a single purpose
  - [ ] This PR only contains configuration changes (package.json, etc.)
  - [ ] This PR only contains code changes (if configuration changes are required, do a separate PR first, then re-base)
- [ ] My code follows the style guidelines of this project, including naming conventions
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] If my code adds or changes components, I have written corresponding stories with Storybook
- [ ] I have performed a self-review of my own code
- [ ] If my code involves visual changes, I am adding applicable screenshots to this thread

### CODE REVIEW

**Comment by julien51** [comment]
Wow, I did not expect you to have the whole thing a single PR ;) I was thinking you'd only work on the low level model stuff!

File: locksmith/src/operations/userOperations.ts [hunk 1]
```diff
@@ -0,0 +1,41 @@
+const ethJsUtil = require('ethereumjs-util')
+const models = require('../models')
+const { User, UserReference } = models
+import * as types from '../types'
+
+class UserOperations {
```
Comment (julien51) [inline locksmith/src/operations/userOperations.ts]: I actually intentionally did not create classes for the other operations. Could you expand on why you think they're required here?

File: locksmith/src/operations/userOperations.ts [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+      recoverySeed: input.recoverySeed,
+      encryptedPrivateKey: input.encryptedPrivateKey,
+    })
+
+    let userReference = await UserReference.create({
+      emailAddress: input.emailAddress.toLowerCase(),
+    })
+
+    if (userReference) {
+      userReference.setUser(user)
```
Comment (julien51) [inline locksmith/src/operations/userOperations.ts]: Could we do that above directly when creating the object? This would save at least on SQL request?

File: locksmith/src/operations/userOperations.ts [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+      userReference.save()
+      return true
+    } else {
+      return false
+    }
+  }
+
+  static async privateKeyLookup(emailAddress: String): Promise<String | null> {
```
Comment (julien51) [inline locksmith/src/operations/userOperations.ts]: `getUserPrivateKeyByEmailAddress` would probably be more consistent with how we named other functions?

File: locksmith/src/operations/userOperations.ts [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+    let user = await UserReference.findOne({
+      where: { emailAddress: emailAddress.toLowerCase() },
+      include: [{ model: User, attributes: ['encryptedPrivateKey'] }],
+    })
+
+    if (user) {
+      return user.User.encryptedPrivateKey
```
Comment (julien51) [inline locksmith/src/operations/userOperations.ts]: Is there no way to do that in a single query?

File: locksmith/src/models/userReference.js [hunk 5]
```diff
@@ -0,0 +1,18 @@
+'use strict'
```
Comment (julien51) [inline locksmith/src/models/userReference.js]: I do not think you need these ;)

File: locksmith/src/models/userReference.js [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+    },
+    {}
+  )
+  UserReference.associate = function(models) {
+    UserReference.belongsTo(models.User, {
+      foreignKey: 'publicKey',
+      targetKey: 'publicKey',
+    })
+  }
+  sequelize.sync()
```
Comment (julien51) [inline locksmith/src/models/userReference.js]: What is this for?

Comment (julien51) [reply locksmith/src/models/userReference.js]: Would you want to add a `UserReference.removeAttribute('id')`?

File: locksmith/src/controllers/userController.ts [hunk 7]
```diff
@@ -0,0 +1,40 @@
+import UserOperations = require('../operations/userOperations')
+import { Request, Response } from 'express-serve-static-core'
+
+class UserController {
```
Comment (julien51) [inline locksmith/src/controllers/userController.ts]: Same here. It looks like we are not using classes for the other controllers. Any reason for the change?

File: locksmith/migrations/20190318152733-create-user.js [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+module.exports = {
+  up: (queryInterface, Sequelize) => {
+    return queryInterface.createTable('Users', {
+      publicKey: {
+        type: Sequelize.STRING,
+        primaryKey: true,
+        allowNull: false,
+        unqiue: true
+      },
+      recoverySeed: {
```
Comment (julien51) [inline locksmith/migrations/20190318152733-create-user.js]: I am not sure `seed` is the right word here. `recoveryPhrase` is probably more accurate? Arguably, a seed is something which "generates" something else. This is used to 'encrypt' the private key.

File: locksmith/migrations/20190318152733-create-user.js [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+        type: Sequelize.STRING,
+        allowNull: false
+      },
+      encryptedPrivateKey: {
```
Comment (julien51) [inline locksmith/migrations/20190318152733-create-user.js]: Let's be more explicit and call it `passwordEncryptedPrivateKey`.

Comment (julien51) [reply locksmith/migrations/20190318152733-create-user.js]: Also, I believe we will use `web3.eth.accounts.encrypt(privateKey, password);` which yields a pretty complex object: https://web3js.readthedocs.io/en/1.0/web3-eth-accounts.html#encrypt

Is a string enough to store that?

File: locksmith/src/operations/userOperations.ts (hunk 1 above)
Comment (akeem) [reply locksmith/src/operations/userOperations.ts]: this was/is me reacting to some typescript things or at least the way that I understand them.  Our js approach written in typescript puts the methods in the global scope, there may be a better way to approach this

File: locksmith/src/operations/userOperations.ts (hunk 2 above)
Comment (akeem) [reply locksmith/src/operations/userOperations.ts]: Agreed, i'll put this and the other call in a transaction. I've started down this direction but paused due to some of the ripples caused

File: locksmith/src/models/userReference.js (hunk 6 above)
Comment (akeem) [reply locksmith/src/models/userReference.js]: this is all part of the sequelize black magic, https://sequelize.readthedocs.io/en/latest/docs/associations/ . From what I understand it glues the models together and provides some of the dynamic methods for convenience

File: locksmith/migrations/20190318152733-create-user.js (hunk 9 above)
Comment (akeem) [reply locksmith/migrations/20190318152733-create-user.js]: Interesting I was thinking that we would use something like HMAC256. Let's discuss the details so that we can get the storage to match our approach

File: locksmith/src/operations/userOperations.ts (hunk 1 above)
Comment (cnasc) [reply locksmith/src/operations/userOperations.ts]: Maybe you're looking for namespaces? https://www.typescriptlang.org/docs/handbook/namespaces.html

Alternatively, just declaring modules.

**Comment by cnasc** [comment]
This looks really good overall. So I can make sure I understand completely, right now this draft API supports 2 operations:
1. Create a user account by POSTing some JSON to the endpoint
2. Retrieve the password-encrypted private key with a GET

I actually have some questions about the second operation, but I think that will be a better discussion in person that we can summarize in text format for the record afterward.

**Comment by julien51** [comment]
@akeem would you be able to start working on splitting this so @cnasc @benwerd and I could work on other related parts?

**Comment by akeem** [comment]
@julien51 how do you envision this being split? At the moment, I am still making adjustments to this but will move this to a non-draft PR once I believe it is in a solid state

**Comment by julien51** [comment]
Split in horizontal layers:
1. the model/migration
2. Operations
3. Controllers

File: locksmith/src/models/userReference.js (hunk 5 above)
Comment (akeem) [reply locksmith/src/models/userReference.js]: this file was generated by Sequelize

File: locksmith/__tests__/operations/userOperations.test.ts [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+
+    it('should normalize the public key/address', async () => {
+      UserReference.create = jest.fn(() => {})
+      expect.assertions(1)
+      await UserOperations.createUser(userCreationDetails)
+      expect(UserReference.create).toHaveBeenCalledWith(
+        {
+          User: {
+            passwordEncryptedPrivateKey: '{"data" : "encryptedPassword"}',
+            publicKey: '0x21cC9C438D9751A3225496F6FD1F1215C7bd5D83',
```
Comment (julien51) [inline locksmith/__tests__/operations/userOperations.test.ts]: Could you actually denormalize (lowercase) the `publicKey` in the `userCreationDetails` so that we know we're actually normalzing it?

File: locksmith/__tests__/operations/userOperations.test.ts [hunk 11]
```diff
@@ -0,0 +1,86 @@
...
+    })
+  })
+
+  describe('when the email address does not exist in persistence', () => {
+    it('returns null', async () => {
+      expect.assertions(1)
+      let result = await UserOperations.getUserPrivateKeyByEmailAddress(
+        'non-existant@example.com'
+      )
+      expect(result).toEqual(null)
```
Comment (julien51) [inline locksmith/__tests__/operations/userOperations.test.ts]: I really think it is important that we actually do not return null but a new encrypted private key so that we are not leaking user data! We should encrypt the email address with a consistent string (some kind of hash of the email address?)

File: locksmith/migrations/20190318152733-create-user.js [hunk 12]
```diff
@@ -0,0 +1,32 @@
...
+        primaryKey: true,
+        allowNull: false,
+        unique: true
+      },
+      recoveryPhrase: {
+        type: Sequelize.STRING,
+        allowNull: false
+      },
+      passwordEncryptedPrivateKey: {
+        type: Sequelize.JSON,
```
Comment (julien51) [inline locksmith/migrations/20190318152733-create-user.js]: that's cool!

File: locksmith/package.json [hunk 13]
```diff
@@ -4,7 +4,7 @@
   "description": "an unlock storage service",
   "main": "server.js",
   "scripts": {
-    "predev": "NODE_ENV=development sequelize db:migrate",
+    "predev": "NODE_ENV=development npm run db:migrate",
```
Comment (julien51) [inline locksmith/package.json]: Could you explain what this changes? Put that into another PR?

File: locksmith/package.json [hunk 14]
```diff
@@ -29,6 +29,7 @@
   "author": "",
   "license": "ISC",
   "dependencies": {
+    "@types/jest": "^24.0.11",
```
Comment (julien51) [inline locksmith/package.json]: Please add dependencies separately!

File: locksmith/package.json [hunk 15]
```diff
@@ -48,6 +49,8 @@
     "sequelize-cli": "^5.4.0",
     "supertest": "^3.3.0",
     "ts-jest": "^24.0.0",
+    "@types/express": "^4.16.1",
```
Comment (julien51) [inline locksmith/package.json]: ditto

File: locksmith/package.json [hunk 16]
```diff
@@ -48,6 +49,8 @@
...
+    "@types/node": "^11.11.3",
```
Comment (julien51) [inline locksmith/package.json]: ditto

File: locksmith/src/controllers/userController.ts [hunk 17]
```diff
@@ -0,0 +1,40 @@
...
+          publicKey: user.publicKey,
+          passwordEncryptedPrivateKey: user.passwordEncryptedPrivateKey,
+          recoveryPhrase: user.recoveryPhrase}
+        )
+
+        let status = creationStatus ? 200 : 400
+        return res.sendStatus(status)
+      }
+    } catch (e) {
+      return res.sendStatus(400)
```
Comment (julien51) [inline locksmith/src/controllers/userController.ts]: 500 would be more appropriate here? Also, I think we should log this?

File: locksmith/src/models/userReference.js (hunk 5 above)
Comment (julien51) [reply locksmith/src/models/userReference.js]: Please remove!

File: locksmith/src/operations/userOperations.ts [hunk 18]
```diff
@@ -0,0 +1,47 @@
...
+const { User, UserReference } = models
+import * as types from '../types'
+
+namespace UserOperations {
+  export async function createUser(
+    input: types.UserCreationInput
+  ): Promise<Boolean> {
+    let userReference = await UserReference.create(
+      {
+        emailAddress: input.emailAddress.toLowerCase(),
```
Comment (julien51) [inline locksmith/src/operations/userOperations.ts]: I don't remember seeing a test where we check with an UPPERCASE email address and we still return the private key corresponding to the lowercase one? Could you add this please?

File: locksmith/src/operations/userOperations.ts [hunk 19]
```diff
@@ -0,0 +1,47 @@
...
+          passwordEncryptedPrivateKey: input.passwordEncryptedPrivateKey,
+        },
+      },
+      {
+        include: User,
+      }
+    )
+
+    if (userReference) {
+      return true
```
Comment (julien51) [inline locksmith/src/operations/userOperations.ts]: I think we can just `return userReference`?

File: locksmith/src/operations/userOperations.ts [hunk 20]
```diff
@@ -0,0 +1,47 @@
...
+  ): Promise<String | null> {
+    let user = await UserReference.findOne({
+      where: { emailAddress: emailAddress.toLowerCase() },
+      include: [{ model: User, attributes: ['passwordEncryptedPrivateKey'] }],
+    })
+
+    if (user) {
+      return user.User.passwordEncryptedPrivateKey
+    } else {
+      return null
```
Comment (julien51) [inline locksmith/src/operations/userOperations.ts]: That's where we should generate a private key and encrypt it with a salted hash of the email address

File: locksmith/src/routes/user.js [hunk 21]
```diff
@@ -0,0 +1,10 @@
+let express = require('express')
+
+let router = express.Router()
+/* eslint-disable import/no-unresolved*/
```
Comment (julien51) [inline locksmith/src/routes/user.js]: Could you explain what this is?

File: locksmith/src/routes/user.js [hunk 22]
```diff
@@ -0,0 +1,10 @@
...
+let userController = require('../controllers/userController')
+
+router.post('/', userController.createUser)
+router.get('/:emailAddress', userController.retrieveEncryptedPrivatekey)
```
Comment (julien51) [inline locksmith/src/routes/user.js]: Could we do `/user/:emailAddress/privatekey` because I think this route is currently too generic and we may create conflicts?

Comment (akeem) [reply locksmith/src/routes/user.js]: It was designed with the idea that /user/<email address> returns the relevant information that we are willing to show about the address passed. In this case the encrypted private key, the idea here being that the data returned could be expanded upon. This can be constricted to only support this specific use case

File: locksmith/src/routes/user.js (hunk 21 above)
Comment (akeem) [reply locksmith/src/routes/user.js]: this is referencing the result of a typescript file that will exist, when build. As there actually isn't a javascript file for this controller prior to being built the linter raises an error

File: locksmith/src/operations/userOperations.ts (hunk 20 above)
Comment (akeem) [reply locksmith/src/operations/userOperations.ts]: This branch of the conditional represents the encrypted private key not existing in storage. From our last conversation, I believe that we were/are not generating private keys at all on the back end and this falls under the concern of the front end

File: locksmith/src/operations/userOperations.ts (hunk 19 above)
Comment (akeem) [reply locksmith/src/operations/userOperations.ts]: At the moment method's return type is `Promise<Boolean>` we could change it to be an any type but I am not sure that it would provide a lot of value to downs stream consumers in this fashion

File: docker/docker-compose.ci.yml [hunk 23]
```diff
@@ -84,7 +84,7 @@ services:
     image: unlock
     user: node
     working_dir: /home/unlock/locksmith
-    command: sh -c './scripts/wait-for.sh db:5432 -- npm run db:migrate && npm run start'
+    command: sh -c './scripts/wait-for.sh db:5432 -- npm run db:migrate && npm run build && node ./build/server.js'
```
[the remark this answers is not in the document]
Comment (akeem) [reply docker/docker-compose.ci.yml]: yep, this is pending a conversation with @cnasc

File: locksmith/src/controllers/userController.ts (hunk 17 above)
Comment (akeem) [reply locksmith/src/controllers/userController.ts]: Can add the logging. Not sure about this being a 500. https://twitter.com/stevelosh/status/372740571749572610?lang=en 

lol

File: locksmith/__tests__/operations/userOperations.test.ts (hunk 11 above)
Comment (akeem) [reply locksmith/__tests__/operations/userOperations.test.ts]: We should have a quick chat about security in general, since we hadn't discussed it in the context of these initial stories. I assumed that we would do it prior to this code went live but probably not in the introduction of this into the code base

File: locksmith/src/operations/userOperations.ts (hunk 19 above)
Comment (julien51) [reply locksmith/src/operations/userOperations.ts]: good point

File: locksmith/src/routes/user.js (hunk 21 above)
Comment (julien51) [reply locksmith/src/routes/user.js]: awww, this is weird. @cnasc is there a way to deal with that in a better way?

File: locksmith/src/routes/user.js (hunk 22 above)
Comment (julien51) [reply locksmith/src/routes/user.js]: I guess than `/user/:emailAddress` is fine, but let's not use `/:emailAddress`.

Comment (julien51) [reply locksmith/src/routes/user.js]: Sorry I misunderstood how the controller prefixes the routes.

**Comment by akeem** [comment]
Closing this as pull requests of the component parts of this have made their way to production
